### PR TITLE
jupyter-widget:  sketch for generic event handling

### PR DESCRIPTION
**Pull request by ibgreen**
For #3864 etc

#### Background

Exploring an approach for adding generic event handling to pydeck.

#### Change List
-

### CODE REVIEW

File: modules/jupyter-widget/src/playground/playground.js [hunk 1]
```diff
@@ -38,6 +38,7 @@ export function initPlayground() {
         container,
         jsonInput: jsonProps,
         tooltip,
+        handleEvent: (name, payload) => sendEventViaTransport(transport, name, payload),
```
Comment (ajduberstein) [inline modules/jupyter-widget/src/playground/playground.js:41]: Python would receive name and payload, name could be passed to a `switch`-like Python statement

File: modules/jupyter-widget/src/playground/playground.js [hunk 2]
```diff
@@ -90,26 +91,33 @@ export function processDataBuffer({binary, convertedJson}) {
   return convertedJson;
 }
 
+// Handles a general event
+function sendEventViaTransport(transport, name, data) {
+  transport.jupyterModel.set('deck_event', JSON.stringify({name, data}));
```
Comment (ajduberstein) [inline modules/jupyter-widget/src/playground/playground.js:96]: Expected to be available within the Python-side of the widget

An observer could be set here, which could trigger a Python function

Comment (ajduberstein) [reply modules/jupyter-widget/src/playground/playground.js:96]: To-do: Figure out how a rapid series of events, like many view state changes, would be received by Python

File: modules/jupyter-widget/src/playground/playground.js [hunk 3]
```diff
@@ -90,26 +91,33 @@ export function processDataBuffer({binary, convertedJson}) {
...
+  transport.jupyterModel.save_changes();
+}
+
 // Handles a click event
+// TODO - integrate as extra processing for click events in sendEventViaTransport
 function handleClick(transport, datum, e) {
   if (!datum || !datum.object) {
-    transport.jupyterModel.set('selected_data', JSON.stringify(''));
+    transport.jupyterModel.set('deck_event', JSON.stringify(''));
```
Comment (ajduberstein) [inline modules/jupyter-widget/src/playground/playground.js:104]: This should be changed back to `selected_data`

File: modules/jupyter-widget/src/playground/create-deck.js [hunk 4]
```diff
@@ -101,40 +101,51 @@ function missingLayers(oldLayers, newLayers) {
   return oldLayers.filter(ol => ol && ol.id && !newLayers.find(nl => nl.id === ol.id));
```
Comment (ajduberstein) [inline modules/jupyter-widget/src/playground/create-deck.js:101]: All comments thus far are notes from a conversation with @ibgreen

File: modules/jupyter-widget/src/playground/playground.js [hunk 5]
```diff
@@ -90,7 +93,27 @@ export function processDataBuffer({binary, convertedJson}) {
...
+    // TODO handle background hover events, for now we'll skip them
+    return;
+  }
+  // TODO Remove circular references without converting to a string
+  const deckEvent = JSON.parse(JSON.stringify({event, data}, filterJsonValue));
+  transport.jupyterModel.send(deckEvent);
+}
+
 // Handles a click event
+// TODO - integrate as extra processing for click events in sendEventViaTransport
```
Comment (ibgreen) [inline modules/jupyter-widget/src/playground/playground.js:116]: Don't forget to plumb through click via the new mechanism as well

File: modules/jupyter-widget/src/playground/create-deck.js [hunk 6]
```diff
@@ -101,40 +101,60 @@ function missingLayers(oldLayers, newLayers) {
...
-) {
+  props,
+  mapboxApiKey,
+  googleMapsKey,
+  handleEvent,
+  getTooltip,
+  container
+}) {
+  // Common deck.gl props for all basemaos
+  const handlers = handleEvent
```
Comment (ajduberstein) [inline modules/jupyter-widget/src/playground/create-deck.js:114]: Skip these handlers in the standalone HTML version of pydeck

**Comment by ajduberstein** [comment]
Added an example notebook and remapped onclick to use the new Jupyter comms. 

Text is dynamically updating on viewport change.

debounce

**Comment by ibgreen** [comment]
> –let me know if you think anything else needs addressing. Otherwise I'll publish a beta later today with these changes.

Please go ahead with the beta. I would like to generalize the back channel from `deck-event` to `transport-event` and handle it within `JupyterTransport.sendMessage()`.

I think this will be a small incremental change at this stage, just make sure the Jupyter model field name is updated correctly on both ends, and (optionally) generalize the code on the python side a bit..

The long term goal is of course to make playground completely abstracted from the jupyter widget, so that we can replace with a `ColabTransport` etc.

**Pull request merged**